core/toolbox/category.js

```diff
@@ -0,0 +1,826 @@
+
+goog.requireType('Blockly.ToolboxItem');
+
+
+/**
+ * Class for a category in a toolbox.
+ * @param {!Blockly.utils.toolbox.CategoryJson} categoryDef The information needed
+ *     to create a category in the toolbox.
+ * @param {!Blockly.IToolbox} toolbox The parent toolbox for the category.
+ * @param {Blockly.ToolboxCategory=} opt_parent The parent category or null if
+ *     the category does not have a parent.
+ * @constructor
+ * @extends {Blockly.CollapsibleToolboxItem}
+ */
+Blockly.ToolboxCategory = function(categoryDef, toolbox, opt_parent) {
+  Blockly.ToolboxCategory.superClass_.constructor.call(
+      this, categoryDef, toolbox);
+
+  /**
+   * The name that will be displayed on the category.
+   * @type {string}
+   * @protected
+   */
+  this.name_ = Blockly.utils.replaceMessageReferences(categoryDef['name']);
+
+  var categories = categoryDef['contents'].filter(function(item) {
+    return item['kind'].toUpperCase() == 'CATEGORY';
+  });
+
+  /**
+   * True if this category has subcategories, false otherwise.
+   * @type {boolean}
+   * @private
+   */
+  this.hasChildren_ = !!categories.length;
+
+  /**
+   * The parent of the category.
+   * @type {?Blockly.ToolboxCategory}
+   * @protected
+   */
+  this.parent_ = opt_parent || null;
+
+  /**
+   * The level that the category is nested at.
+   * @type {number}
+   * @protected
+   */
+  this.level_ = this.parent_ ? this.parent_.getLevel() + 1 : 0;
+
+  /**
+   * The colour of the category.
+   * @type {string}
+   * @protected
+   */
+  this.colour_ = this.getColour_(categoryDef);
+
+  /**
+   * The html container for the category.
+   * @type {?Element}
+   * @protected
+   */
+  this.htmlDiv_ = null;
+
+  /**
+   * The html element for the category row.
+   * @type {?Element}
+   * @protected
+   */
+  this.rowDiv_ = null;
+
+  /**
+   * The html elmeent that holds children of the category row.
+   * @type {?Element}
+   * @protected
+   */
+  this.rowContents_ = null;
+
+  /**
+   * The html element for the toolbox icon.
+   * @type {?Element}
+   * @protected
+   */
+  this.iconDom_ = null;
+
+  /**
+   * Container for any children categories.
+   * @type {?Element}
+   * @protected
+   */
+  this.subcategoriesDiv_ = null;
+
+  /**
+   * All the css class names that are used to create a category.
+   * @type {!Blockly.ToolboxCategory.CssConfig}
+   * @protected
+   */
+  this.cssConfig_ = {
+    'container': 'blocklyToolboxCategory',
+    'row': 'blocklyTreeRow',
+    'rowContentContainer': 'blocklyTreeRowContentContainer',
+    'icon': 'blocklyTreeIcon',
+    'label': 'blocklyTreeLabel',
+    'contents': 'blocklyToolboxContents',
+    'selected': 'blocklyTreeSelected',
+    'openIcon': 'blocklyTreeIconOpen',
+    'closedIcon': 'blocklyTreeIconClosed',
+  };
+
+  var cssConfig = categoryDef['cssconfig'] || categoryDef['cssConfig'];
+  Blockly.utils.object.mixin(this.cssConfig_, cssConfig);
+
+  /**
+   * Whether or not the category should display its children.
```

Comment:
Are these children the subcategories, or the blocks?

Reply:
Update here: #4230